packages/next/src/server/app-render.tsx

```diff
@@ -1476,10 +1477,11 @@ export async function renderToHTMLOrFlight(
       renderResult: RenderResult
     ): Promise<string> => {
       const renderChunks: string[] = []
+      const textDecoder = new TextDecoder()
 
       const writable = {
         write(chunk: any) {
-          renderChunks.push(decodeText(chunk))
+          renderChunks.push(decodeText(chunk, textDecoder))
         },
         end() {},
         destroy() {},
```

Comment:
We need to perform a final call to the `textDecoder` here to flush any trailing bytes that have not completed. This only comes up when an incomplete octet set is encountered at the end of of the stream:

```js
const chunk = new TextEncoder().encode('abc💩').slice(0, -1);

const decoder = new TextDecoder();
let output = decoder.decode(chunk, { stream: true });

console.assert(output === 'abc');

// Flush the decoder to receive the final REPLACEMENT CHARACTER
// for the unfinished emoji
output += decoder.decode(new Uint8Array(0), { stream: false });

console.assert(output === 'abc�');

Reply:
Yep
I've mentioned that here https://github.com/vercel/next.js/pull/46564#issuecomment-1448833198

I've guessed no one cares.
I can make another PR for all six usages if you need it

Reply:
Thanks for the example, will follow up later!